[webkit.UncountedLambdaCapturesChecker] Ignore trivial functions and [[clang::noescape]].

rniwa opened the pull request:
This PR makes webkit.UncountedLambdaCapturesChecker ignore trivial functions as well as the one being passed to an argument with [[clang::noescape]] attribute. This dramatically reduces the false positive rate for this checker.

To do this, this PR replaces VisitLambdaExpr in favor of checking lambdas via VisitDeclRefExpr and VisitCallExpr. The idea is that if a lambda is defined but never called or stored somewhere, then capturing whatever variable in such a lambda is harmless.

VisitCallExpr explicitly looks for direct invocation of lambdas and registers its DeclRefExpr to be ignored in VisitDeclRefExpr. If a lambda is being passed to a function, it checks whether its argument is annotated with [[clang::noescape]]. If it's not annotated such, it checks captures for their safety.

Because WTF::switchOn could not be annotated with [[clang::noescape]] as function type parameters are variadic template function so we hard-code this function into the checker.

In order to check whether "this" pointer is ref-counted type or not, we override TraverseDecl and record the most recent method's declaration.

In addition, this PR fixes a bug in isUnsafePtr that it was erroneously checking whether std::nullopt was returned by isUncounted and isUnchecked as opposed to the actual boolean value.

Finally, this PR also converts the accompanying test to use -verify and adds a bunch of tests.

rniwa committed:
[webkit.UncountedLambdaCapturesChecker] Ignore trivial functions and [[clang::noescape]].

This PR makes webkit.UncountedLambdaCapturesChecker ignore trivial
functions as well as the one being passed to an argument with
[[clang::noescape]] attribute. This dramatically reduces the false
positive rate for this checker.

To do this, this PR replaces VisitLambdaExpr in favor of checking
lambdas via VisitDeclRefExpr and VisitCallExpr. The idea is that if a
lambda is defined but never called or stored somewhere, then capturing
whatever variable in such a lambda is harmless.

VisitCallExpr explicitly looks for direct invocation of lambdas and
registers its DeclRefExpr to be ignored in VisitDeclRefExpr. If a lambda
is being passed to a function, it checks whether its argument is
annotated with [[clang::noescape]]. If it's not annotated such, it
checks captures for their safety.

Because WTF::switchOn could not be annotated with [[clang::noescape]] as
function type parameters are variadic template function so we hard-code
this function into the checker.

In order to check whether "this" pointer is ref-counted type or not,
we override TraverseDecl and record the most recent method's declaration.

In addition, this PR fixes a bug in isUnsafePtr that it was erroneously
checking whether std::nullopt was returned by isUncounted and isUnchecked
as opposed to the actual boolean value.

Finally, this PR also converts the accompanying test to use -verify and
adds a bunch of tests.

haoNoQ reviewed (commented):
The attribute gets some action! Nice!!

I noticed that currently the attribute's documentation doesn't say it can be placed on lambdas. But it doesn't look like it's actively rejected either. So it might be a good idea to update the documentation as part of the patch. Possibly even have a bit of a broader discussion about this - though I don't really see any downsides to simply allowing this.

rniwa committed:
Merge branch 'main' into trivial-lambdas-and-noescape-param

rniwa commented:
> I noticed that currently the attribute's documentation doesn't say it can be placed on lambdas. But it doesn't look like it's actively rejected either. So it might be a good idea to update the documentation as part of the patch. Possibly even have a bit of a broader discussion about this - though I don't really see any downsides to simply allowing this.

I'm more than happy to bring the discussion up but which venue (forum, mailing list, etc...) should I be using to solicit a broader discussion?

rniwa commented on clang/lib/StaticAnalyzer/Checkers/WebKit/UncountedLambdaCapturesChecker.cpp:
```diff
@@ -45,32 +52,119 @@ class UncountedLambdaCapturesChecker
+        return true;
+      }
+
+      // WTF::switchOn(T, F... f) is a variadic template function and couldn't
+      // be annotated with NOESCAPE. We hard code it here to workaround that.
```
I see. It's probably an overkill for this one function but I'd keep that in my mind.

rniwa committed:
Override TraverseCXXMethodDecl to simplify the code per review comment.
Also fix a nullptr crash.

rniwa committed:
Merge branch 'main' into trivial-lambdas-and-noescape-param

t-rasmud commented on clang/lib/StaticAnalyzer/Checkers/WebKit/UncountedLambdaCapturesChecker.cpp:
```diff
@@ -45,32 +52,118 @@ class UncountedLambdaCapturesChecker
+          unsigned ArgIndex = 0;
+          for (auto *Param : Callee->parameters()) {
+            if (ArgIndex >= CE->getNumArgs())
+              break;
```
nit: Could we just `return true` instead of a `break` here?

t-rasmud commented on clang/lib/StaticAnalyzer/Checkers/WebKit/UncountedLambdaCapturesChecker.cpp:
```diff
@@ -89,7 +182,25 @@ class UncountedLambdaCapturesChecker
+      Os << "Implicitly captured ";
+    }
+
+    Os << "raw-pointer 'this' to ref-counted / CheckedPtr capable type is "
```
nit: Can this warning be better worded? `.. ref-counted (or CheckedPtr capable) type ...`? Or anything else that conveys that it's unsafe to capture raw-pointer to a ref-counted type or a "CheckedPtr capable" type.

rniwa replied:
Rephrased to say "ref-counted type or CheckedPtr-capable type"

t-rasmud reviewed (approved):
LGTM! Left a couple nits, mainly about readability of the warning.

rniwa committed:
Address the review comments.

Also fix a bug that TraverseCXXMethodDecl was erroneously calling TraverseDecl,
and cleanup the test.
